pkgs/development/compilers/flutter/flutter.nix

```diff
@@ -55,37 +71,67 @@ let
       export SNAPSHOT_PATH="$FLUTTER_ROOT/bin/cache/flutter_tools.snapshot"
       export STAMP_PATH="$FLUTTER_ROOT/bin/cache/flutter_tools.stamp"
 
-      export DART_SDK_PATH="${dart}"
+      export DART_SDK_PATH="$(which dart)"
 
-      HOME=../.. # required for pub upgrade --offline, ~/.pub-cache
-                 # path is relative otherwise it's replaced by /build/flutter
+      ${if usePreload then ''
+
+      # we need to make the home $out here because dart pub get and pub cache preload will, place dependencies, which
+      # may also be required later at runtime of the flutter command, in $HOME/.pub-cache
+      mkdir -p $out
+      HOME=$out
+
+      dart pub cache preload .pub-preload-cache/* || (echo "dart pub cache preload failed. If you are overriding flutter's sources to build an older version, set usePreload to false in mkFlutter (preload is needed for dart >= 2.19 and flutter >= 3.7.9)" && exit 1)'' else ''
+      HOME=../..
+      ''}
 
       pushd "$FLUTTER_TOOLS_DIR"
-      ${dart}/bin/dart pub get --offline
+        rm -rf test
+        dart pub get --offline -v
       popd
 
       local revision="$(cd "$FLUTTER_ROOT"; git rev-parse HEAD)"
-      ${dart}/bin/dart --snapshot="$SNAPSHOT_PATH" --packages="$FLUTTER_TOOLS_DIR/.dart_tool/package_config.json" "$SCRIPT_PATH"
+      dart --snapshot="$SNAPSHOT_PATH" --packages="$FLUTTER_TOOLS_DIR/.dart_tool/package_config.json" "$SCRIPT_PATH"
       echo "$revision" > "$STAMP_PATH"
       echo -n "${version}" > version
 
-      rm -r bin/cache/{artifacts,dart-sdk,downloads}
-      rm bin/cache/*.stamp
+      rm -r bin/cache/dart-sdk
     '';
 
+    nativeBuildInputs = [
+      autoPatchelfHook
+    ];
+
     installPhase = ''
       runHook preInstall
 
-      mkdir -p $out
       cp -r . $out
       mkdir -p $out/bin/cache/
       ln -sf ${dart} $out/bin/cache/dart-sdk
 
       runHook postInstall
     '';
 
+    postFixup = ''
+      # flutter is really senstive about the executable being called flutter and being in the
+      # bin folder, so this is easier than creating another bin directory somewhere and mkWrapper-ing.
+
+      sed -i '2i\
+      export PUB_CACHE=\''${PUB_CACHE:-"\$HOME/.pub-cache"}\
+      export ANDROID_EMULATOR_USE_SYSTEM_LIBS=1\
+      export PATH=$PATH:${lib.makeBinPath [
+        bash
+        curl
+        dart
+        git
+        unzip
+        which
+        xz
+      ]}
+      ' $out/bin/flutter
+    '';
+
     doInstallCheck = true;
-    nativeInstallCheckInputs = [ which ];
+    installCheckInputs = [ which git ];
```

Comment:
These binaries are to be executed on the build platform in install check phase, they belong to `nativeInstallCheckInputs`